doc/proposals/dualstack.md

```diff
@@ -0,0 +1,280 @@
+      DNS                 cnitypes.DNS      `json:"dns"`
+      Range               string            `json:"range"`
+      RangeStart          net.IP            `json:"range_start,omitempty"`
+      RangeEnd            net.IP            `json:"range_end,omitempty"`
+      GatewayStr          string            `json:"gateway"`
+      EtcdHost            string            `json:"etcd_host,omitempty"`
+      EtcdUsername        string            `json:"etcd_username,omitempty"`
+      EtcdPassword        string            `json:"etcd_password,omitempty"`
+      EtcdKeyFile         string            `json:"etcd_key_file,omitempty"`
+      EtcdCertFile        string            `json:"etcd_cert_file,omitempty"`
+      EtcdCACertFile      string            `json:"etcd_ca_cert_file,omitempty"`
+      LeaderLeaseDuration int               `json:"leader_lease_duration,omitempty"`
+      LeaderRenewDeadline int               `json:"leader_renew_deadline,omitempty"`
+      LeaderRetryPeriod   int               `json:"leader_retry_period,omitempty"`
+      LogFile             string            `json:"log_file"`
+      LogLevel            string            `json:"log_level"`
+      OverlappingRanges   bool              `json:"enable_overlapping_ranges,omitempty"`
+      SleepForRace        int               `json:"sleep_for_race,omitempty"`
+      Gateway             net.IP
+      Kubernetes          KubernetesConfig `json:"kubernetes,omitempty"`
+      ConfigurationPath   string           `json:"configuration_path"`
+      PodName             string
+      PodNamespace        string
+ }
+```
+
+Corresponding changes will also be required in `whereabouts/pkg/allocate/allocate.go`, `whereabouts/pkg/config/config.go` etc.
+
+<hr>
+
+## Alternative Design <a name="alternative"></a>
+
+Alternatively, if we only want to support DualStack (i.e. at max one IPv4 and one IPv6 address), we can have a simpler config change which is easily backward compatible. _(But I still prefer the first approach even though it is requires more effort)_
+
+We can have change in IPAM Config as below.
+
+<table>
+<tr>
+<th>Old IPAM Config</th>
+<th>Changes</th>
+<th>New IPAM Config</th>
+</tr>
+<tr>
+<td>
+  
+```json
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28",
+        "exclude": [
+           "192.168.2.229/30",
+           "192.168.2.236/32"
+        ]
+      }
+}
+```
+  
+</td>
+<td>
+
+```diff
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28",
++       "secondary_range": "2001::0/116",
+        "exclude": [
+           "192.168.2.229/30",
+           "192.168.2.236/32",
++          "2001::0/124"
```

Comment:
how would you map the entries on the exclusion list to the correct range ? Here it actually seems doable since the ranges are from different IP families, but, what if you want to use this feature to configure multiple IPs in the same interface ?

Reply:
We can take separate unions of IPv4 and IPv6 excluded ranges and just skip the IPs that belongs to that union set/list :)
I don't think we need to map it to the exact range. ^^